### DISCUSSION TITLE
Center Screen

### DISCUSSION THREAD

**Discussion by HideakiAtsuyo** [emilk/egui#1801] [Q&A]
Hello, how can I make the window show at center screen when I launch the program instead of anywhere ?

**Comment by willser** [comment]
I find a way to make window show at center.
`Code`
````rust
#[derive(Deserialize, Serialize, Default)]
pub struct Weaver {}

impl App for Weaver {
    fn update(&mut self, ctx: &Context, frame: &mut Frame) {}

    fn persist_native_window(&self) -> bool {
        false
    }
}

fn main() {
    let mut native_options = eframe::NativeOptions::default();
    let mut monitors = winit::event_loop::EventLoop::new().available_monitors();
    let first_monitor_size = monitors.next().unwrap().size();
    let width = 300.0;
    let height = 100.0;
    native_options.initial_window_size = Some(egui::Vec2::new(width, height));
    native_options.initial_window_pos = Some(egui::Pos2::new(
        (first_monitor_size.width as f32 - width) / 2.0,
        (first_monitor_size.height as f32 - height) / 2.0,
    ));
    eframe::run_native(
        "weaver2",
        native_options,
        Box::new(|creation_context| Box::new(Weaver::default())),
    );
}


````
`Cargo.toml`
````toml
winit = "0.26.1"
````

**Reply by HideakiAtsuyo** [reply]
That's working well yeah(in my dreams), I always used the center to avoid the windows spawn anywhere it looks bad in my opinion

**Reply by HideakiAtsuyo** [reply]
Actual result with this:
![](https://i.imgur.com/2mM6EU9.gif)

**Reply by willser** [reply]
Try clear storage or rename your app.

**Reply by willser** [reply]
Following those code,I guess you need clear old app storage or rename your app.

https://github.com/emilk/egui/blob/e76c919c7e70c208c9a6209b9fe3369e7b6db99d/eframe/src/native/run.rs#L57-L64

https://github.com/emilk/egui/blob/e76c919c7e70c208c9a6209b9fe3369e7b6db99d/eframe/src/native/epi_integration.rs#L72-L86

I test again,it's work fine in my pc.I suggest to rename your app for test.

https://user-images.githubusercontent.com/64425242/178269462-3dea4178-64ec-41ad-907c-91385521ff5a.mp4

**Comment by relu91** [comment]
btw with the latest winit you don't have the monitor information until you run the event loop. Would this mean that it is now impossible to center the window before the creation? Does it make sense to instantiate briefly the event loop just to retrieve the screen information needed? 

See https://github.com/rust-windowing/winit/issues/3723

**Comment by relu91** [comment]
I ended up doing this trick inside the update function. I don't really like it... if someone has better options please let me know:
```rust
        let mut desired_pos: Option<Vec2>= None;
        ctx.input(|input| {
            if let Some(size) = input.viewport().monitor_size {
                let center = pos2(size.x/2.0, size.y /2.0) - pos2(310.0, 120.0) ;
                desired_pos = input.viewport().outer_rect.and_then(|current_position| {
                    if current_position.min.x.round() == center.x && current_position.min.y.round() == center.y {
                        None
                    } else {
                        Some(center)
                    }
                })
            }
        });

        if let Some(pos) = desired_pos {
            let move_command = ViewportCommand::OuterPosition(pos2(pos.x, pos.y));
            ctx.send_viewport_cmd(move_command);
        }
```

**Reply by Tyarel8** [reply]
If you are using `eframe` the `NativeOptions` config has a [centered](https://docs.rs/eframe/latest/eframe/struct.NativeOptions.html#structfield.centered) field that centers the window in your screen before creating it.

**Reply by relu91** [reply]
It has always been there in plain sight 🫠🤦‍♂️thank you for pointing it out.

**Reply by seyoatda** [reply]
> If you are using `eframe` the `NativeOptions` config has a [centered](https://docs.rs/eframe/latest/eframe/struct.NativeOptions.html#structfield.centered) field that centers the window in your screen before creating it.

If I want to set the window for a precise position, is there any cleaner plan? Or this is the only way to implement it trickily?😢
